### PR TITLE
ipn/ipnlocal: use syspolicy to determine collection of posture data 

### PR DESCRIPTION
**Pull request by kradalby**
This PR will allow users/orgs to instruct the client to collect Posture Checking data (currently serial numbers) via MDM settings in the Registry on Windows and defaults on macOS.

Updates #5902

Signed-off-by: Kristoffer Dalby <kristoffer@tailscale.com>

### CODE REVIEW

File: ipn/ipnlocal/c2n.go [hunk 1]
```diff
@@ -222,9 +223,11 @@ func (b *LocalBackend) handleC2NPostureIdentityGet(w http.ResponseWriter, r *htt
...
 	res := tailcfg.C2NPostureIdentityResponse{}
 
-	// TODO(kradalby): Use syspolicy + envknob to allow Win registry,
-	// macOS defaults and env to override this setting.
-	if b.Prefs().PostureChecking() {
+	// Only collect serial numbers if enabled on the client,
+	// this will first check syspolicy, MDM settings like Registry
+	// on Windows or defaults on macOS. If they are not set, it falls
+	// back to the cli-flag, `--posture-checking`.
+	if enabled, _ := syspolicy.GetBoolean(syspolicy.PostureChecking, b.Prefs().PostureChecking()); enabled {
```
Comment (clairew) [inline ipn/ipnlocal/c2n.go:230]: GetBoolean is tested but can you test how your own error handling will work w/ syspolicy?

Comment (kradalby) [reply ipn/ipnlocal/c2n.go]: I changed it to log the error and return the default, thanks.

Comment (kradalby) [reply ipn/ipnlocal/c2n.go]: I changed it to log the error and return the default, thanks.

File: ipn/ipnlocal/c2n.go [hunk 2]
```diff
@@ -229,9 +230,21 @@ func (b *LocalBackend) handleC2NPostureIdentityGet(w http.ResponseWriter, r *htt
...
+	// back to the cli-flag, `--posture-checking`.
+	enabled, err := syspolicy.GetBoolean(syspolicy.PostureChecking, b.Prefs().PostureChecking())
+	if err != nil {
+		b.logf(
+			"c2n: failed to read PostureChecking from syspolicy, returning default from cli: %s",
+			err,
+		)
+
+		enabled = b.Prefs().PostureChecking()
+	}
```
[the remark this answers is not in the document]
Comment (knyar) [reply ipn/ipnlocal/c2n.go:245]: In this case I think logging is fine – I suspect in most cases these errors will actionable to the users, but not to us, and client logs are probably the most obvious place to find them.

In general my thinking is that we would only want to return errors as part of a response if we have a way to show them to users in Admin Console or aggregate and present in some other way. Until we want to do that, I think logging non-critical errors on the client will be sufficient, and critical errors we can return to control in a c2n response (like we do on line 250 below).

File: ipn/ipnlocal/c2n.go [hunk 3]
```diff
@@ -229,9 +230,21 @@ func (b *LocalBackend) handleC2NPostureIdentityGet(w http.ResponseWriter, r *htt
...
+	// on Windows or defaults on macOS. If they are not set, it falls
+	// back to the cli-flag, `--posture-checking`.
+	enabled, err := syspolicy.GetBoolean(syspolicy.PostureChecking, b.Prefs().PostureChecking())
+	if err != nil {
+		b.logf(
+			"c2n: failed to read PostureChecking from syspolicy, returning default from cli: %s",
+			err,
+		)
+
+		enabled = b.Prefs().PostureChecking()
```
Comment (knyar) [inline ipn/ipnlocal/c2n.go:244]: It seems like `%s` might refer to the "default from cli", but it's actually the error message. Perhaps printing both would be more clear? For example:

```suggestion
		enabled = b.Prefs().PostureChecking()
		b.logf("c2n: failed to read PostureChecking from syspolicy, returning default from CLI: %s; got error: %s", enabled, err)
```

**Review by knyar — APPROVED** [review]

**Pull request merged**